### PR TITLE
 socket_zep: don't always listen on local port

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

- if no local address is specified, don't listen on a port, just connect to the remote address
- ~~sent a dummy packet at startup (that is dropped by `socket_zep`) to announce one's presence if a UDP dispatcher is used~~

### Testing procedure

`socket_zep` will not bind to a local port unless explicitly requested to do.

### Issues/PRs references

split off #14755

### CODE REVIEW

File: cpu/native/socket_zep/socket_zep.c [hunk 1]
```diff
@@ -350,26 +350,70 @@ void socket_zep_setup(socket_zep_t *dev, const socket_zep_params_t *params)
...
+        dev->sock_fd = socket(AF_INET6, SOCK_DGRAM, 0);
+    } else {
+        dev->sock_fd = res;
+    }
+
+    _connect_remote(dev, params);
+
+    /* send dummy data to connect to dispatcher */
+    /* (no valid ZEP header, will by dropped in _recv() */
+    real_write(dev->sock_fd, "HELLO", 5);
```
Comment (miri64) [inline cpu/native/socket_zep/socket_zep.c:415]: Even though this is not "standard compliant" to ZEP (I am not aware of a standard though and my implementation is based on the [wireshark dissector code](https://github.com/boundary/wireshark/blob/master/epan/dissectors/packet-zep.c#L25-L43) and what little the [wireshark doc provides](https://www.wireshark.org/docs/dfref/z/zep.html)) I would prefer if this packet is recognizable as a ZEP packet. For ZEPv2 there are [two types known: ACK and DATA](https://github.com/boundary/wireshark/blob/master/epan/dissectors/packet-zep.c). However, the ZEP type field is one byte, so enough to space to invent other types.

How about defining (driver internally, not in the global `sys/include/net/zep.h` header) a new type

```C
#define SOCKET_ZEP_V2_TYPE_HELLO   (255)
```

The packet could be the same format as DATA (as the wireshark dissector [always falls back to DATA when the packet is not an ACK](https://github.com/boundary/wireshark/blob/07eade8124fd1d5386161591b52e177ee6ea849f/epan/dissectors/packet-zep.c#L150-L158)) but leave all fields 0, except for the `reserved` field which could contain the "HELLO" string.

This way the packet is disected in a wireshark dump as a ZEP packet, and also clearly visable as a hello packet.

**Review by miri64 — CHANGES_REQUESTED** [review]
I am fine with the "don't always listen on local port" part of this PR. However, the second change you "hid" in here, I'd like to discuss (probably would have been better to make those rather unrelated changes two PRs).

File: cpu/native/socket_zep/socket_zep.c (hunk 1 above)
Comment (benpicco) [reply cpu/native/socket_zep/socket_zep.c]: Ok, dropped it from this PR.

My reasoning was that the packet should be easy to distinguish as non-ZEP packet and dropped early.

But I can also do the version with the ZEP header if you think that's cleaner.

**Comment by miri64** [comment]
The usage message needs to be adapted:
```
    -z [<laddr>:<lport>,]<raddr>:<rport> --zep=[<laddr>:<lport>,]<raddr>:<rport>
        provide a ZEP interface with local address and port (<laddr>, <lport>)
        and remote address and port (default local: [::]:17754).
        Required to be provided SOCKET_ZEP_MAX times
```

**Comment by miri64** [comment]
With this however, the link layer address of the node is not stable anymore :-/ (the IID is generated from the (laddr,lport,raddress,rport) tuple).

**Comment by miri64** [comment]
> With this however, the link layer address of the node is not stable anymore :-/ (the IID is generated from the (laddr,lport,raddress,rport) tuple).

(not blocking though, as their is a solution for that: https://github.com/RIOT-OS/RIOT/pull/14758)

File: cpu/native/startup.c [hunk 2]
```diff
@@ -314,8 +314,8 @@ void usage_exit(int status)
 #endif
 #if defined(MODULE_SOCKET_ZEP) && (SOCKET_ZEP_MAX > 0)
 "    -z [<laddr>:<lport>,]<raddr>:<rport> --zep=[<laddr>:<lport>,]<raddr>:<rport>\n"
-"        provide a ZEP interface with local address and port (<laddr>, <lport>)\n"
-"        and remote address and port (default local: [::]:17754).\n"
+"        provide a ZEP interface with (optional) local address and port (<laddr>, <lport>).\n"
+"        The ZEP interface will connect to the remote address and optionally listen on a local address.\n"
```
Comment (miri64) [inline cpu/native/startup.c:318]: Line-length...

Comment (miri64) [reply cpu/native/startup.c]: Patch changed, but lines are still longer than 80 chars and can be divided up ;-).

File: cpu/native/socket_zep/socket_zep.c (hunk 1 above)
Comment (miri64) [reply cpu/native/socket_zep/socket_zep.c]: IMHO it makes the purpose of this packet clearer if just looking at a dump when it is a ZEP packet.

File: cpu/native/startup.c (hunk 2 above)
Comment (benpicco) [reply cpu/native/startup.c]: I though 80 characters was only a recommendation…

Comment (miri64) [reply cpu/native/startup.c]: Only if you can't find a way around it, is my interpretation of that convention. Let's not make it worse, especially in terminal output. Right now, the output is broken on a small (size as opened by Gnome) terminal window (IIRC the other broken output about GPIO is also yours ;-)):

image

Comment (miri64) [reply cpu/native/startup.c]: > (The other lines in this file are also > 80 chars, just right above this line)

They are in a huge minority btw. Disregarding this addition there are

https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L310-L314 (should be fixed)

https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L417 (a link, so no way around it)

The remaining are only overlapping a few characters. Could be fixed, but only if you are bone-headed about it:

https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L420-L421
https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L433
https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L455
https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L501
https://github.com/RIOT-OS/RIOT/blob/870c8920bd1b9125f2807c58a5e963fb4b0c6bce/cpu/native/startup.c#L608

Comment (benpicco) [reply cpu/native/startup.c]: Ok, should fit now.

**Review by miri64 — APPROVED** [review]
ACK.

<details><summary><code>examples/gnrc_networking</code> still works as expected, when using the correct ports</summary>

##### Terminal 1

```console
$ TERMFLAGS="-z [::]:17754" make -C examples/gnrc_networking -j term
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking'
/home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking/bin/native/gnrc_networking.elf -z [::]:17754  
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2021.01-devel-461-g22ba7-HEAD)
RIOT network stack example application
All up, running the shell now
> ifconfig
ifconfig
Iface  7  HWaddr: DE:73  Channel: 26  NID: 0x23 
          Long HWaddr: 00:5A:45:50:0A:00:DE:73 
          L2-PDU:102  MTU:1280  HL:64  RTR  
          6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::25a:4550:a00:de73  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff00:de73
          
          Statistics for Layer 2
            RX packets 2  bytes 86
            TX packets 5 (Multicast: 5)  bytes 129
            TX succeeded 6 errors 0
          Statistics for IPv6
            RX packets 2  bytes 128
            TX packets 5 (Multicast: 5)  bytes 320
            TX succeeded 5 errors 0

> 
```

##### Terminal 2
```console
$ netstat -up
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
udp6       0      0 localhost:56946         localhost:zep           ESTABLISHED 300818/gnrc_network 
$ TERMFLAGS="-z [::]:17754,[::1]:56946" make -C examples/gnrc_networking term
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking'
/home/mlenders/Repositories/RIOT-OS/RIOT/examples/gnrc_networking/bin/native/gnrc_networking.elf -z [::]:17754,[::1]:56946  
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2021.01-devel-461-g22ba7-HEAD)
RIOT network stack example application
All up, running the shell now
> ping6 fe80::25a:4550:a00:de73
ping6 fe80::25a:4550:a00:de73
12 bytes from fe80::25a:4550:a00:de73%7: icmp_seq=0 ttl=64 rssi=255 dBm time=1.745 ms
12 bytes from fe80::25a:4550:a00:de73%7: icmp_seq=1 ttl=64 rssi=255 dBm time=1.699 ms
12 bytes from fe80::25a:4550:a00:de73%7: icmp_seq=2 ttl=64 rssi=255 dBm time=1.697 ms

--- fe80::25a:4550:a00:de73 PING statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 1.697/1.713/1.745 ms
> ^C
native: exiting

```

</details>

**Comment by miri64** [comment]
> `examples/gnrc_networking` still works as expected, when using the correct ports

(should have mentioned: I replaced `gnrc_netdev_default` with `socket_zep` in the application's Makefile ;-))

**Pull request merged**